Cross-Shard Congestion Control (#539)
near/NEPs: neps/nep-0539.md

for a transaction from when it is accepted to when it finishes execution becomes
larger and larger, as the receipts need to queue behind those already in the
system.

First attempts to limit the memory consumption have been added without protocol
changes. This is known as "local congestion control" and can be summarized in
two rules:

- Limit the transaction pool to 100 MB.
  https://github.com/near/nearcore/pull/9172
- Once we have accumulated more than 20k delayed receipts in a shard,
  chunk-producers for that shard stop accepting more transactions.
  https://github.com/near/nearcore/pull/9222

But this does not put any limits on what shards will accept for other shards.
For example, when a particular contract is popular, the contract's shard will
eventually stop accepting new transactions but all other shards keep accepting
more and more transactions that produce receipts for the popular contract.
Therefore the number of delayed receipts keeps growing indefinitely.

Cross-shard congestion control addresses this issue by stopping new transactions
at the source and delaying receipt forwarding when the receiving shard has
reached its memory limit.

## Specification

> The proposal adds fields to chunks headers, it adds a trie columns, it changes

Reviewer: ```suggestion
The proposal adds fields to chunks headers, it adds a new trie column, it changes
```

the rules to select transactions, and it changes the chunk execution flow. The

> next four sections specifies each of these changes in more details.

Reviewer: ```suggestion
next four sections specify each of these changes in more details.
```

### Chunk header changes

We change the chunk header to include congestion information.
Specifically, we include two indicators.

```rust
ShardChunkHeaderInnerV3 {
  // ...all field from ShardChunkHeaderInnerV2
  incoming_congestion: u16
  general_congestion: u16
}
```

This adds 4 bytes to the chunk header, increasing it from 374 bytes to 378
bytes in borsh representation. (Assuming no validator proposals included.)
This in turn increases the block size by 4 bytes per shard.

> While the numbers are stored as `u16`,for the remainder of the NEP

Reviewer: ```suggestion
While the numbers are stored as `u16`, for the remainder of the NEP
```

specification, `incoming_congestion` and `general_congestion` values are
interpreted as numbers between 0 and 1, dividing the stored `u16` by `u16::MAX`
(65535).

### Changes to transaction selection

Today, transactions are taken from the chunk producer's pool until `tx_limit` is
reached, where `tx_limit` is computed as follow.

```python
# old
tx_limit = 500 Tgas if len(delayed_receipts) < 20_000 else 0
```

We replace the formula for the transaction limit to depend on the
`general_congestion` computed in the previous shard variable (between 0 and 1):

```python
# new

> MIN_GAS = 5 Tgas

Reviewer 2: Should it be 5 Tgas or 0 Tgas?

Author: I think 5 Tgas is better but it's up for debate.

The idea here is that we allow a small amount of gas for accepting new transactions to other shard, if we know those other shards have `incoming_congestion = 0`. (see condition further below)

The thing is, if we allow it to go to zero, we allow complete starvation of a shard by other shards.
In practical terms, if my accounts happens to be on the same shard as HOT and everyone uses HOT, I can no longer do anything on any shard with my account, because I need to start at my own shard. Even if I only spend 2 Tgas on function call conversion on the congested shard and then 302 Tgas on a different shard which is currently idle.

Also, the harm of including 5 Tgas seems small since we still have 995 Tgas left for working on receipts to reduce the congestion.

Reviewer 2: I see, makes sense.

MAX_GAS = 500 Tgas
tx_limit = mix(MAX_GAS, MIN_GAS, general_congestion)
```

*TODO: fine-tune `MIN_GAS` and `MAX_GAS`*

In the pseudo code above, we borrow the [`mix`](https://docs.gl/sl4/mix)
function from GLSL for linear interpolation.

> `mix(x, y, a)`
> 
> `mix` performs a linear interpolation between x and y using a to weight between
> them. The return value is computed as $x \times (1 - a) + y \times a$. 

> Plus, we add the additional rule to reject all transactions to shard with an

Author: Kind of yes, in the proposal we are only rejecting it at the chunk producer level. Everything else is not a protocol change and will not make it into this NEP. So we don't really define what should be treated as accepted or not accepcted.

Your described ideal world can certainly be implemented with the proposed spec here. If we take the vocabulary from [issue #6837](https://github.com/near/nearcore/issues/6837), anything except wait_until=None should not return success before a transaction is actually in a block. And the rejection filter happens before inclusion in a block.

In the most naive solution, we wait until the chunk producer rejects it and don't tell the RPC node waiting for it. This would be bad and the transaction just keeps waiting until some timeout.

Middle-ground: We signal back to the RPC node that the tx was rejected. This can be tricky since we forward it to several chunk producers and maybe some of them accept while others reject, if congestion is just building up or just being resolved. But we can just wait for all of them to answer before telling the user that it was rejected.

Dream scenario: RPC nodes run the rejection filter before even forwarding and reject it instantly. This also reduced load on the network to chunk producers.

> Plus, we add the additional rule to reject all transactions to shard with an

Reviewer 3: sounds good.  Agreed that such discussions are outside the scope of the NEP and should definitely be possible with this NEP.

`incoming_congestion` above a certain threshold.

```python
def filter_tx(tx):
  INCOMING_CONGESTION_THRESHOLD = 0.5
  if incoming_congestion(tx.receiver_shard_id) > INCOMING_CONGESTION_THRESHOLD
    tx.reject()
  else
    tx.accept()
```

> *TODO: fine-tune `INCOMING_CONGESTION_THRESHOLD`*

Reviewer 4: We likely need a lower one if maximum is 100Pgas. From a user experience point of view, I think over 10Pgas of delay is already quite suboptimal and should trigger the congestion control mechanism

    - Subtract `receipt.gas()` from `outgoing_gas_limit[receipt.receiver]` for
      each receipt drained.
    - Keep receipts in the buffer if the gas limit would be negative.
    - Add the removed receipts to the outgoing receipts of the new chunk.
3. Convert all transactions to receipts included in the chunk.
    - Local receipts, which are receipts where the sender account id is equal to
      the receiver id, are set aside as local receipts for step 4.
    - Non-local receipts up to `outgoing_gas_limit[receipt.receiver]` for the
      respective shard go to the outgoing receipts list of the chunk.
    - (new) Non-local receipts above `outgoing_gas_limit[receipt.receiver]` for
      the respective shard go to the outgoing receipts buffer.
4. Execute receipts in the order of `local`, `delayed`, `incoming`.
    - Don't stop before all receipts are executed or more than 1000 Tgas have
      been burnt. Burnt gas includes the burnt gas from step 3.
    - Outgoing receipts up to what is left in
      `outgoing_gas_limit[receipt.receiver]` per shard (after step 3) go to the
      outgoing receipts list of the chunk.

>     - (new) Above `outgoing_gas_limit[receipt.receiver]` receipts go to the
>       outgoing buffer.

Reviewer: ```suggestion
    - (new) Outgoing receipts above `outgoing_gas_limit[receipt.receiver]`
      go to the outgoing receipts buffer.
```

5. Remaining local or incoming receipts are added to the end of the `delayed`
   receipts queue.
6. (new) Compute own congestion information for the next block.:
    ```python
    MAX_CONGESTION_INCOMING_GAS = 100 PGas
    gas_backlog = sum([receipt.gas() for receipt in delayed_receipts_queue])
    incoming_congestion = gas_backlog / MAX_CONGESTION_INCOMING_GAS
    incoming_congestion = min(1.0, incoming_congestion)

    MAX_CONGESTION_MEMORY_CONSUMPTION = 500 MB
    memory_consumption = 0
    memory_consumption += sum([receipt.size() for receipt in delayed_receipts_queue])
    memory_consumption += sum([receipt.size() for receipt in postponed_receipts_queue])
    memory_consumption += sum([receipt.size() for receipt in outgoing_receipts_buffer])
  
    general_congestion = memory_consumption / MAX_CONGESTION_MEMORY_CONSUMPTION
    general_congestion = min(1.0, general_congestion)
    ```

*TODO: fine-tune `MAX_CONGESTION_INCOMING_GAS` and `MAX_CONGESTION_MEMORY_CONSUMPTION`*

> In the formula above, we receipt gas and size are defined as:

Reviewer: ```suggestion
In the formula above, the receipt gas and the receipt size are defined as:
```

```python
def gas(receipt):
  return receipt.attached_gas + receipt.exec_gas

def size(receipt):
  return len(borsh(receipt))
```

### Change to Trie

We store the outgoing buffered receipts in the trie, similar to delayed receipts
but in their own separate column. Therefore we add a trie column

> `DELAYED_RECEIPT_OR_INDICES: u8 = 13;`. Then we read and write analogue to tje

Reviewer: ```suggestion
`BUFFERED_RECEIPT_OR_INDICES: u8 = 13;`. Then we read and write analogue to the
```

Is this the name you wanted to go for? Anyway it shouldn't be the same as for delayed receipts :)

> `BUFFERED_RECEIPT_OR_INDICES: u8 = 13;`. Then we read and write analogue to the

Reviewer 5: Minor detail: I know we use this pattern for `DELAYED_RECEIPT_OR_INDICES`, but it seems to be that way for historical reasons ([see commit message here](https://github.com/near/nearcore/commit/7365af44942f984a23ae7c1956d9b96802e52c9b)).

For this new queue it would be clearer to have separate `BUFFERED_RECEIPT` and `BUFFERED_RECEIPT_INDICES` columns.

      each receipt drained.
    - Keep receipts in the buffer if the gas limit would be negative.
    - Subtract `receipt.gas()` from `outgoing_congestion` and `receipt.size()` from `receipt_bytes` for the local shard for every forwarded receipt.
    - Add the removed receipts to the outgoing receipts of the new chunk.
4. Convert all transactions to receipts included in the chunk.
    - Local receipts, which are receipts where the sender account id is equal to
      the receiver id, are set aside as local receipts for step 4.
    - Non-local receipts up to `outgoing_gas_limit[receipt.receiver]` for the
      respective shard go to the outgoing receipts list of the chunk.
    - (new) Non-local receipts above `outgoing_gas_limit[receipt.receiver]` for
      the respective shard go to the outgoing receipts buffer.
    - (new) For each receipt added to the outgoing buffer, add `receipt.gas()`
      to `outgoing_congestion` and `receipt.size()` to `receipt_bytes` for
      the local shard.
5. Execute receipts in the order of `local`, `delayed`, `incoming`, `yield-resume time-out`.
    - Don't stop before all receipts are executed or more than 1000 Tgas have

>       been burnt. Burnt gas includes the burnt gas from step 3.

Author: This should be step 4, thanks for pointing it out. Forgot to change the reference when I added step 1.

---
NEP: 539
Title: Cross-Shard Congestion Control
Authors: Waclaw Banasik <waclaw@near.org>, Jakob Meier <inbox@jakobmeier.ch>
Status: New
DiscussionsTo: https://github.com/nearprotocol/neps/pull/539
Type: Protocol
Version: 1.0.0
Created: 2024-03-21
LastUpdated: 2024-04-21
---

## Summary

We propose to limit the outgoing receipts included in each chunk. Limits depend
on the congestion level of that receiving shard and are to be enforced by the
protocol.

Congestion is primarily tracked in terms of the gas of all delayed receipts.
Chunk producers must ensure they stop accepting new transactions if the receiver
account is on a shard with a full delayed receipts queue.

Forwarding of receipts that are not directly produced by a transaction, namely
cross-contract calls and delegated receipts, is limited by the receivers overall
congestion level. This includes the gas of delayed receipts and the gas of
receipts that have not been forwarded, yet, due to congestion control
restrictions. Additionally, the memory consumption of both receipt types can
also cause congestion.

> This proposal replaces the local congestion control rules already in place.

Reviewer 6: Nit: I think it supercedes/extends the existing rules rather than fully replacing them (we still have transaction pool limit and delayed receipt queue limit).

Author: True, I've added a few words to describe what we keep and what we replace.

For example, when a particular contract is popular, the contract's shard will
eventually stop accepting new transactions but all other shards keep accepting
more and more transactions that produce receipts for the popular contract.
Therefore the number of delayed receipts keeps growing indefinitely.

Cross-shard congestion control addresses this issue by stopping new transactions
at the source and delaying receipt forwarding when the receiving shard has
reached its congestion limit.

## Specification

The proposal adds fields to chunks headers, it adds a new trie column, it changes
the rules to select transactions, and it changes the chunk execution flow. The
next four sections specify each of these changes in more details.

### Chunk header changes

We change the chunk header to include congestion information.
Specifically, we include two indicators.

```rust
ShardChunkHeaderInnerV3 {
  // ...all fields from ShardChunkHeaderInnerV2

  congestion_info: CongestionInfo
}
```

Where alternatives for congestion info are decribed in more detail below in the 
`Efficiently computing the congestion information` section. 

> Depending on the alternative solution this adds between 5 and 48 bytes to the

Author: Good idea. I moved it to the alternative sections, making it concrete in the specification part of the NEP.

chunk header, increasing it from 374 bytes up to 422 bytes in borsh representation.
(Assuming no validator proposals included.) This in turn increases the block size
by up to 48 bytes per shard.

Including all this information in the chunk header enables efficient validation.
Using either the previous chunk header or the state, and the list of receipts 
applied and forwarded, a validator can check that the congestion rules described 
in this NEP are fulfilled.

### Changes to transaction selection

Transaction selection is changed to reject new transactions when the system is
congested, to reduce to total workload in the system.

Today, transactions are taken from the chunk producer's pool until `tx_limit` is
reached, where `tx_limit` is computed as follow.

```python
# old
tx_limit = 500 Tgas if len(delayed_receipts) < 20_000 else 0
```

We replace the formula for the transaction limit to depend on the

> `incoming_congestion` computed in the previous shard variable (between 0 and 1):

Reviewer 6: Do you mean computed in the previous chunk for this shard?

Author: Yes. The sentence should be fixed now to make more sense.

```python
# new
MIN_TX_GAS = 20 Tgas
MAX_TX_GAS = 500 Tgas

> tx_limit = mix(MAX_TX_GAS, MIN_TX_GAS, incoming_congestion)

Author: Yes that makes a lot of sense. I've done so now, please take another look.

backlog of delayed receipts.

In the pseudo code above, we borrow the [`mix`](https://docs.gl/sl4/mix)
function from GLSL for linear interpolation.

> `mix(x, y, a)`
> 
> `mix` performs a linear interpolation between x and y using a to weight between
> them. The return value is computed as $x \times (1 - a) + y \times a$. 

But more importantly, we add a more targeted rule to reject all transactions to
a shard with a congestion level above a certain threshold.

```python
def filter_tx(tx):
  REJECT_TX_CONGESTION_THRESHOLD = 0.25
  if congestion(tx.receiver_shard_id) > REJECT_TX_CONGESTION_THRESHOLD
    tx.reject()
  else
    tx.accept()
```

Here, `congestion()` is the maximum of incoming, outgoing, and memory congestion.

This stops (some) new incoming work at the source for when a shard is using too
much memory to store unprocessed receipts, or there is already too much work
piled up for that shard.

Chunk validators must validate that the two rules above are respected in a
produced chunk.

### Changes to chunk execution

We change chunk execution to hold back receipt forwarding to congested shards.
This has two effects. 

> 1. It spreads the memory consumption from the congested shard to the shards

Author: Yeah I like that wording much better :) Done as suggested.

   producing the excess workload.
2. When user demand is consistently higher than the system can handle, this
   mechanism lets backpressure propagate shard-by-shard until the shards
   responsible for accepting too many receipts are reached and transaction
   filtering kicks in.

To accomplish this, we add 3 new steps to chunk execution (enumerate as 1, 2, 6
below) and modify how outgoing receipts are treated in the transaction
conversion step (3) and in the receipts execution step (4).

> The new chunk execution then follows this order.

Author: Sorry, there were too many rounds of renaming things and in the end it was rather inconsistent. I completely missed that, thanks for pointing it out. It should be fixed now.

    // First inclusive index in the queue.
    pub first_index: u64,
    // Exclusive end index of the queue
    pub next_available_index: u64,
}
```

The `BUFFERED_RECEIPT` column stores receipts keyed by 
`TrieKey::BufferedReceipt{ receiving_shard: ShardId, index: u64 }`.

The `BufferedReceiptIndices` map defines which queues exist, which changes
during resharding. For each existing queue, all receipts in the range
`[first_index, next_available_index)` (inclusive start, exclusive end) must
exist under the key with the corresponding shard.

### Notes on parameter fine-tuning

Below are the reasons why each parameter is set to the specific value given above.

For more details, a spreadsheet with the full analysis can be found here:
https://docs.google.com/spreadsheets/d/1Vt_-sgMdX1ncYleikYY8uFID_aG9RaqJOqaVMLQ37tQ/edit#gid=0

#### Queue sizes

The parameters are chosen to strike a balance between guarantees for short
queues and utilization. 20 PGas delayed receipts means that incoming receipts
have to wait at most 20 chunks to be applied. And it can guarantee 100%
utilization as long as the ratio between burnt and attached gas in receipts is

> above 1 to 20.

Author: No, we don't enforce it in our proposal.

It is an interesting idea to investigate. But note that some patterns inevitably burn less than 1/20 in the first step. A delegate action, for example, burns very little gas on the relayer's shard. Any enforcement measures needs to consider what happens to such use cases.

> above 1 to 20.

Reviewer 6: Got it.

True, this would be quite tricky to get right with multi-hop transactions.

      acceptable.

## Future possibilities

While this proposal treats all requests the same, it sets the base for a proper
transaction priority implementation. We co-designed this proposal with
[NEP-541](https://github.com/near/NEPs/pull/541), which adds a transaction
priority fee. On a very high level, the fee is used to auction off a part of the
available gas per chunk to the highest bidders.

We also expect that this proposal alone will not be the final solution for
congestion control. Rather, we just want to build a solid foundation in this NEP
and allow future optimization to take place on top of it.

For example, estimations on how much gas is burnt on each shard could help with
better load distribution in some cases.

We also forsee that the round-robin scheduling of shard allowed to forward even
under full congestion is not perfect. It is a key feature to make deadlocks
provably impossible, since every shard is guaranteed to make a minimum progress
after N rounds. But it could be beneficial to allocate more bandwidth to shards
that actually have something to forward, or perhaps it would be better to stop

> forwarding anything for a while. The current proposal allows chunk producers to

Reviewer 6: > The current proposal allows chunk producers to experiment with this without a protocol version change.

True, but practically speaking, chunk producers will use the default implementation in nearcore. So we need to make sure that the signals that we provide (e.g. `allowed_shard`) are enough to iterate to a reasonable policy that deals with the congestion pattern seen on mainnet today.

In my understanding, round-robin on average introduces N/2 = 3 blocks of latency (where N=6 is the number of shards) to cross-contract calls. This could be noticeable for some large contracts like `sweat` and `sweat.claim` that have a lot of communication and are located on different shards that are often congested.

I wonder if there is some way to make the behavior of `allowed_shard` smoother (e.g. allowing multiple senders in each round).

Reviewer 6: Actually, maybe under congestion, the 3 blocks delay is not that big of a deal on top of 20 blocks delay from the delayed receipts queue. Do simulations that you ran provide any insights here?

Author: Yes, we do think round-robin is the best choice in the short-term. We just also want to give the protocol team the ability to experiment with it and change it, once we see how congestion control manifests itself on mainnet.

There are main ideas why we think round-robin is best.

1. This is a last-resort measure. The linear reduction of forwarding limits is designed to prevent it from happening in the first place. If we hit it, we are badly congested.
2. If we hit this round-robin because of incoming congestion, we don't introduce extra latency. Anything forwarded is waiting behind a long delayed queue anyway, it might as well wait in outgoing buffers.
3. If we hit it because of outgoing or memory congestion, we are in trouble and are much more concerned about deadlock prevention than about minimizing latencies. The only alternative solution we know of that does not deadlock is a global transaction stop. But global-transaction stop leads to terrible fairness problems, where a single heavily used shard affects all shards and can even cause them to become congested as well due to reduced capacity on them.
4. Allowing multiple shards to send at once quickly leads to conflicts with the 300 Tgas max gas value of a receipt. We went with the simplest possible solution: a single allowed shard.

Author: > Actually, maybe under congestion, the 3 blocks delay is not that big of a deal on top of 20 blocks delay from the delayed receipts queue. Do simulations that you ran provide any insights here?

As far as simulations go, it's kind of hard to get into this phase and stay in it. Once in round-robin, it often goes in-and-out of it every few blocks.

"Relayed hot" in the spreadsheet is probably the best example, which has some extra graphs at the bottom of the document. Using the limits we have in the proposal, the median transaction delay is at 19 blocks, the 90th percentile at 24 blocks. This includes 2 cross-shard hops but I'm pretty sure only one is operating under round-robin from time to time. So yeah, 3 blocks isn't a huge deal.

experiment with this without a protocol version change.

Lastly, a future optimization could do better transaction rejection for meta
transactions. Instead of looking only at the outer transaction receiver, we
could also look at the receiver of the delegate action, which is most likely
where most gas is going to be burnt, and use this for transaction rejection.

## Consequences

### 

### Positive

- Accepted transaction have lower latencies compared to today under congestion.

> - Storage and memory requirements on validator for storing receipts are bounded.

Reviewer 6: I don't understand why outgoing receipts buffers are bounded in size. Or at least I only understand why this is a case under the assumption that `filter_tx` above is working effectively. Otherwise, I can imagine a case when we keep accepting and processing transactions to shard X (due to a low value of `incoming_congestion`) that generate an outgoing receipt (as a cross-contract call) to a congested shard Y which can't all be processed timely and accumulate in unbounded fashion.

Author: Absolutely right, outgoing buffers are not hard-bounded to the memory size parameter we give it. We should probably clarify what we mean by a bound.

In simulations, we tried to push it past the soft-limit as much as possible. One example is in the spreadsheet. You can see 1.3 GB memory usage for the "Big Linear Imbalance" workload, with the limit at 1GB. This workload has 500 B / 60 Tgas receipts that indirectly produce 2 MB / 10 Tgas receipts.

But we tried with many other workloads, too. Every workload will fill up the system with some amount of receipts before it locks down new traffic from entering. And max memory usage then essentially depends on what memory the receipts in the system use to finish enough (child) receipts to allow more traffic to enter again.

This arguing only works if the system actually locks down every workload eventually. This is hard, if not impossible, to prove for general dynamic workloads. That's why we decided to simulate as much as we could.

In practice, we can still cause the memory requirements to 3-4x the limit. But all simulated workloads stabilized around that or lower. In comparison, today's "strategy" would lead to >500GB memory usage within 2000 blocks even for the aforementioned "Big Linear Imbalance" workload.

All that said, we are aware that for malicious workloads, the max in bytes could still be larger than we can handle. There are a couple malicious ideas we never tested. But from the beginning, we put malicious workloads as second priority behind getting a basic congestion control for regular users in place.

### Neutral

> - More transactions are rejected at the chunk producer level.

Author: This is an important topic and a great point to be raised. Let me describe my thinking, which also reflects what we discussed in the last weekly congestion control meeting.

From a protocol point of view, we add a condition to reject transactions, meaning it would be invalid to add them to a chunk. What happens in the transaction pool isn't really relevant on that level, hence unspecified, or simply implementation choices.

The question regarding transaction pool queueing depends on at least two implementation choices.

- Are transactions removed from the transaction pool if they are rejected once due to congestion?
- Are RPC nodes filtering transaction according to current congestion before forwarding them to chunk producers?

Personally, I hope we can go for "Yes and Yes". RPC nodes should only accept a transaction if it has a high probability of being accepted. In case of rejection, users get immediate feedback. It's up to wallets to resend as often as appropriate. And chunk-producers should aggressively remove transaction that can't be included in chunks at the moment. For good UX, this rejection should also propagate back to the user through the RPC nodes. But if it's rare enough, we might get away without this in the initial deployment.

If we implement is as "No and No," then you are exactly right, we will generally queue more in the transaction pool until its size limit is reached.

"Yes and No" is also possible, which keeps the transaction pool short. But here it is even more important to propagate the rejection information back to the user, or else we have terrible UX since RPC nodes keep waiting for inclusion of the tx in a chunk when it was already dropped. But this back propagation is non-trivial to implement, given that multiple chunk producers are involved, so we might have to ship it later than the base congestion control mechanism.

"No and Yes" is also worth considering. Probably the easiest implementation without breaking UX. It's just odd because RPC nodes essentially control how large the transaction pool grows, rather than the chunk-producers that actually store it. But it could be a stepping stone towards "Yes and Yes".

Which ever way we go, I believe we don't need to change anything on the protocol level. And rejected transaction handling is fundamentally a client-side question that I believe should not be defined on the protocol level. In this discussion here, we only need to ensure the specification allows for decent client implementations.

---
NEP: 539
Title: Cross-Shard Congestion Control
Authors: Waclaw Banasik <waclaw@near.org>, Jakob Meier <inbox@jakobmeier.ch>
Status: New
DiscussionsTo: https://github.com/nearprotocol/neps/pull/539
Type: Protocol
Version: 1.0.0
Created: 2024-03-21
LastUpdated: 2024-05-15
---

## Summary

We propose to limit the transactions and outgoing receipts included in each
chunk. Limits depend on the congestion level of that receiving shard and are to
be enforced by the protocol.

Congestion is primarily tracked in terms of the total gas of all delayed receipts.
Chunk producers must ensure they stop accepting new transactions if the receiver
account is on a shard with a full delayed receipts queue.

Forwarding of receipts that are not directly produced by a transaction, namely
cross-contract calls and delegated receipts, is limited by the receiver's overall
congestion level. This includes the gas of delayed receipts and the gas of
receipts that have not been forwarded, yet, due to congestion control
restrictions. Additionally, the memory consumption of both receipt types can
also cause congestion.

This proposal extends the local congestion control rules already in place. It

> keeps the transaction pool size limit as it but replaces the old delayed receipt

Reviewer 7: ```suggestion
keeps the transaction pool size limit as is but replaces the old delayed receipt
```

the delayed receipt queue, which itself is stored in the state of the trie.
There is exactly one delayed receipts queue per shard.

**Outgoing receipts** are all newly produced receipts as a result of executing
receipts or when converting a transaction to non-local receipts. In the absence
of congestion, they are all stored in the output of applying receipts, as a
simple list of receipts.

The **outgoing receipts buffer** is a data structure added by cross-shard
congestion control. Each shard has one instance of it in the state trie for
every other shard. We use this buffer to store outgoing receipts temporarily
when reaching receipt forwarding limits.

**Receipt limits** are measured in gas and size. Gas in this context refers to
the maximum amount of gas that could be burn when applying the receipt. The
receipt size is how much space it takes in memory, measured in bytes.

The **congestion level** is an indicator between 0 and 1 that determines how
strong congestion prevention measures should be. The maximum congestion measured
is reached at congestion level 1. This value is defined separately for each
shard and computed as the maximum value of the following congestion types.

> - **Incoming congestion** increases linearly with the amount of gas is in

Reviewer 7: ```suggestion
- **Incoming congestion** increases linearly with the amount of gas in
```

      each receipt drained.
    - Keep receipts in the buffer if the gas limit would be negative.
    - Subtract `receipt.gas()` from `outgoing_congestion` and `receipt.size()`
      from `receipt_bytes` for the local shard for every forwarded receipt.
    - Add the removed receipts to the outgoing receipts of the new chunk.
4. Convert all transactions to receipts included in the chunk.
    - Local receipts, which are receipts where the sender account id is equal to
      the receiver id, are set aside as local receipts for step 4.
    - Non-local receipts up to `outgoing_gas_limit[receipt.receiver]` for the
      respective shard go to the outgoing receipts list of the chunk.
    - (new) Non-local receipts above `outgoing_gas_limit[receipt.receiver]` for
      the respective shard go to the outgoing receipts buffer.
    - (new) For each receipt added to the outgoing buffer, add `receipt.gas()`
      to `outgoing_congestion` and `receipt.size()` to `receipt_bytes` for
      the local shard.
5. Execute receipts in the order of `local`, `delayed`, `incoming`, `yield-resume time-out`.
    - Don't stop before all receipts are executed or more than 1000 Tgas have

>       been burnt. Burnt gas includes the burnt gas from step 4.

Author: Step 4 converts transaction to receipts. The send gas of the created receipts and all inner actions are burnt in that step.
